rfc/20250317-ephemeral-resources.md

```diff
@@ -0,0 +1,646 @@
+
+### Outputs
+
+For enabling ephemeral outputs, these are the basic steps that need to be taken:
+* Update config to support the `ephemeral` attribute.
+* Mark the outputs with a new mark and ensure that the marks are propagated correctly.
+  * We should use boolean marks, as no additional information is required to be carried. When introducing the marks for these, extra care should be taken in *all* the places marks are handled and ensure that the existing implementation around marks is not affected.
+* Based on the marks, ensure that the output cannot be used in other contexts than the ephemeral ones (see the User Documentation section for more details on where this is allowed).
+
+> [!TIP]
+>
+> For an example on how to properly introduce a new mark in the outputs, you can check the [PR](https://github.com/opentofu/opentofu/pull/2633) for the deprecated outputs.
+
+Strict rules:
+* A root module cannot define ephemeral outputs since are useless.
+* Any output that wants to use an ephemeral value, it needs to be marked as ephemeral too. Otherwise, it needs to show an error:
+   ```hcl
+    │ Error: Output not marked as ephemeral
+    │
+    │   on mod/main.tf line 33, in output "password":
+    │   33:   value = reference.to.ephemeral.value
+    │
+    │ In order to allow this output to store ephemeral values add `ephemeral = true` attribute to it.
+   ```
+* Any output from a root module that is referencing a write only attribute needs to be marked as sensitive too. Otherwise, an error should be raised
+  ```hcl
+    │ Error: An output referencing a sensitive value needs to be marked with sensitive too
+    │
+    │   on main.tf line 32:
+    │   32: output "write_only_out" {
+    │
+    │ For security reasons, OpenTofu requires any output that is referencing a sensitive value to also be configured the same. If the root module really wants to export this sensitive value, you need to annotate it with the following argument:
+    │     sensitive = true
+  ```
+
+Considering the rules above, root modules cannot have any ephemeral outputs defined.
+
+### Locals
+Any `local` declaration will be marked as ephemeral if in the expression that initialises it an ephemeral value is used:
+```hcl
+variable "var1" {
+  type = string
+}
+
+variable "var2" {
+  type = string
+}
+
+variable "var3" {
+  type      = string
+  ephemeral = true
+}
+
+locals {
+  eg1 = var.var1 == "" ? var.var2 : var.var1 // not ephemeral
+  eg2 = var.var2 // not ephemeral
+  eg3 = var.var3 == "" ? var.var2 : var.var1 // ephemeral because of var3 conditional
+  eg4 = var.var1 == "" ? var.var2 : var.var3 // ephemeral because of var3 usage
+  eg5 = "${var.var3}-${var.var1}" // ephemeral because of var3 usage
+  eg6 = local.eg4 // ephemeral because of eg4 is ephemeral
+}
+```
+
+Once a local is marked as ephemeral, this can be used only in other ephemeral contexts. Check the `User Documentation` section for more details on the allowed contexts.
+
```

Comment:
let's not forget to document here that variables that take in ephemeral marked values should also be implicitly marked as ephemeral

Reply:
Added information in 11a29cb050fd95ab55b5f6d5df01bba3c73af854.
A variable will not become ephemeral strictly from referencing an ephemeral value. In order for it to be able to work with ephemeral values, it needs to be configured specifically.